### PR TITLE
[DUOS-2414][risk=no] Use match v3 for all match results

### PR DESCRIPTION
**Pull request by fboulnois**
### Addresses

https://broadworkbench.atlassian.net/browse/DUOS-2414

### Summary

Use V3 match for all match results.

### Notes

I originally tried to propagate the `APPROVE`, `DENY`, and `ABSTAIN` case across the codebase, but this added much more code to the codebase and importantly required me to change the `matchentity` column in the database from a boolean to an enum. This had additional consequences including requiring us to migrate all the current matches which made the problem more challenging.

As a result I picked a simpler approach which instead adds a boolean `abstain` column in the database and use the `APPROVE`, `DENY`, `ABSTAIN` where it makes sense, without propagating it to the whole codebase.

----
Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

- Label PR with a Jira ticket number and include a link to the ticket
- Label PR with a security risk modifier [no, low, medium, high]
- PR describes scope of changes
- Get a minimum of one thumbs worth of review, preferably two if enough team members are available
- Get PO sign-off for all non-trivial UI or workflow changes
- Verify all tests go green
- Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

**Review by kaitlynoflaherty — APPROVED** [review]
Looks good!

File: src/main/java/org/broadinstitute/consent/http/service/MatchService.java [hunk 1]
```diff
@@ -142,17 +144,17 @@ public List<Match> createMatchesForConsent(String consentId) {
...
+                    logWarn("Error finding matches for consent: " + consentId);
+                    matches.add(matchFailure(consentId, dar.getReferenceId(), List.of()));
                 }
             });
         });
         return matches;
     }
 
-    private Match singleEntitiesMatchV2(Dataset dataset, DataAccessRequest dar) {
+    private Match singleEntitiesMatchV3(Dataset dataset, DataAccessRequest dar) {
```
Comment (connorlbark) [inline src/main/java/org/broadinstitute/consent/http/service/MatchService.java]: nit: should we add a test for this function or to `matchSuccess` to ensure that the logic for abstain/approve/deny is mapping to the right booleans?

**Review by connorlbark — APPROVED** [review]
looks great! i could see it with a few more simple tests to round it out / become more confident it all works, but it looks there and the abstain column is a great approach for keeping the update simple to integrate and to understand.

File: src/main/java/org/broadinstitute/consent/http/service/MatchService.java (hunk 1 above)
Comment (fboulnois) [reply src/main/java/org/broadinstitute/consent/http/service/MatchService.java]: I added tests to have complete coverage of `singleEntitiesMatchV3`.

**Review by connorlbark — APPROVED** [review]

File: src/main/java/org/broadinstitute/consent/http/models/Match.java [hunk 2]
```diff
@@ -135,4 +140,12 @@ public void addFailureReason(String reason) {
...
         }
     }
+
+    public static Match matchFailure(String consentId, String purposeId, List<String> failureReasons) {
+        return new Match(consentId, purposeId, false, false, true, MatchAlgorithm.V3, failureReasons);
+    }
+
+    public static Match matchSuccess(String consentId, String purposeId, DataUseMatchResultType match, List<String> failureReasons) {
+        return new Match(consentId, purposeId, Approve(match), Abstain(match), false, MatchAlgorithm.V3, failureReasons);
+    }
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/models/Match.java:150]: In theory, failure reasons should always be an empty list in the success case.

**Review by rushtong — APPROVED** [review]
Looks good 👍🏽 - see optional comment inline.

File: src/main/java/org/broadinstitute/consent/http/models/Match.java (hunk 2 above)
Comment (fboulnois) [reply src/main/java/org/broadinstitute/consent/http/models/Match.java:150]: The failure reasons are extracted from the [response entity](https://github.com/DataBiosphere/consent/pull/1997/files#diff-b2fbd6a813de88461f1c1800da590a32cf7ac7312857be8a6b5566509240f799R174), so from what I can tell it is possible to have a MatchResult with [various associated reasons](https://github.com/DataBiosphere/consent-ontology/blob/develop/src/main/java/org/broadinstitute/dsde/consent/ontology/datause/DataUseMatcherV3.java#L75), and these reasons will be sent as [failure reasons](https://github.com/DataBiosphere/consent-ontology/blob/develop/src/main/java/org/broadinstitute/dsde/consent/ontology/resources/MatchResource.java#L97) in the json.

So although the POST is successful and the `consent` Match does not have the boolean `failed` set to true, it still may have one or more "failure reasons".

**Pull request merged**